[PhpUnitBridge] Added a CoverageListener to enhance the code coverage report

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/8416

---

The code coverage computed by PHPUnit is not very accurate by default as it marks a line as tested as soon as it has been executed.
For example, if you have two classes A and B where A is using B and you write test only for the class A then the class B will be marked as tested.

You can fix this issue by adding `@covers A` on top of the class ATest, but it's a bit boring.

This Listener add this annotation on each test if it's applicable:
* If an annotation already exists, we do nothing.
* We try to find the SUT thanks to the Test class name, if it does not exist, we do nothing

---

If you wan to see it in action: https://github.com/lyrixx/phpunit-auto-cover

---

The PR is not finished, I think we could add this listener to symfony itself.

What do you think?

lyrixx committed:
[PhpUnitBridge] Added a CoverageListener to enhance the code coverage report

stof commented:
@lyrixx is there a way to opt-out the guessing for some test cases (smoke tests used to cover things through functional tests for instance) ?

nicolas-grekas commented:
I like it :)

lyrixx commented:
@stof As soon as you add a annotation on your test class, the listener adds nothing.
Thus, if it's really needed, we could make this listener configurable.

dunglas commented:
Nice!

lyrixx commented:
This PR did not get so much attraction. IMHO, it's nice. Should I finished it or it will be a waste of time?
(This is really easy, as we have to deal with many PHPUnit version (PSR4 vs Pear))

Simperfit commented:
IMO It should be finished, I may use it ;)

sebastianbergmann commented:
A test listener should only listen. In the future, you will only be able to listen.

The proper solution to this problem is to follow best practices and set `forceCoversAnnotation="true"` in your PHPUnit configuration.

stof commented:
@sebastianbergmann well, this listener is an attempt at guessing the covered class based on a convention instead of forcing to put the annotation everywhere. What would be the proper way to implement such convention-based coverage configuration in PHPUnit ?

fabpot commented:
My first reaction was: why is it related to Symfony? Why isn't it something part of PHPUnit itself it that's something useful?

lyrixx commented:
@fabpot that's a good question. I guess I created the PR in Symfony because I contribute more to Symfony than to PHPUnit. And we already have some listener in Symfony.

medinae commented:
Useful feature 👍 I agree with @lyrixx as others similar listeners are existing in SF.

fabpot commented:
@lyrixx Can you add some information in the documentation?

lyrixx commented:
@fabpot Keep in mind this PR is really not ready. I have to make it work for different PHPUnit version.

anyway, what documentation are you referring to? (local PHP doc or the symfony/symfony-docs repo?)

lyrixx commented:
I updated the PR to support PHPUnit 5.* and 6.*.

Locally, I also updated all phpunit.xml.dist to add the listener, but the phpunit bridge is not required. Should I update them too?

keradus commented:
IMO, if we already try to figure out what is the expected class to be tested, we could already put it into code and do it only _once_, commit and forget about it. no need to do it all the times on runtime.
Then, it could be watched to have the annotation for new tests, eg with suggestion of @sebastianbergmann  and PHP CS Fixer/fabbot `php_unit_test_class_requires_covers` rule

nicolas-grekas commented:
Would it be possible to add a test, by mean of a phpt for example? this would ensure that we can detect whenever phpunit changes its internals and breaks the feat.
Since this is purely opt-in, I'm personally OK to accept it. With a doc PR :)

keradus commented:
@nicolas-grekas , why do it in runtime instead of actually store it as annotation at all test files, as I suggest ?

nicolas-grekas commented:
@keradus the purpose of this PR is exactly to leverage a convention in order to *not* have to add `@cover` to all test files... because it's boring to repeat yourself in each and every test case I guess? :)

keradus commented:
explicit is better than implicit ;)
on top of that, if automate guess wrongly, you don't know (as you don't even see what was chosen) and collect data not how you wish to. if for some reason you wish to have some non-standard cover restriction in given test, guessing on the fly won't help.

lyrixx commented:
> if automate guess wrongly,

how is it possible?

keradus commented:
> how is it possible?

it could as there is no tests, so if one day there will be some newer autoloader spec, dir structure or anything, it could silently fail matching src class for given test class

> and BTW this feature is opt-in ; So if you don't like it, don't use it ;)

of course, but still I think it's nice to think about other possible solutions.
on one side you have logic hidden, on other you have it explicit

nicolas-grekas commented:
Would be great to allow injecting a different convention, by providing the listener with a callable (typically a static method configured in phpunit.xml)

lyrixx commented:
> Would be great to allow injecting a different convention, by providing the listener with a callable (typically a static method configured in phpunit.xml)

Done. You can see an implementation [here](https://github.com/lyrixx/phpunit-auto-cover/commit/c732eb73863401526faf6cdb93060972c61dc252)

> The callable could be used per-method/test case btw?

You mean configurable? if so => I don't think so. Or maybe with an annotation. But frankly, I think It's overkill.

nicolas-grekas commented on src/Symfony/Bridge/PhpUnit/CoverageListener.php:
```diff
@@ -0,0 +1,42 @@
+     */
+    class CoverageListener extends BaseTestListener
+    {
+        public function __construct(callable $sutFqcnResolver = null)
```
typo, should be "zut" :)
what does "sut" mean?

lyrixx replied:
System Under Test. It is used many times on https://phpunit.de/manual/6.3/en/test-doubles.html

nicolas-grekas commented:
would need a doc PR, and a CHANGELOG entry I guess

keradus commented on src/Symfony/Bridge/PhpUnit/CoverageListener.php:
```diff
@@ -0,0 +1,42 @@
+// Using an early return instead of a else does not work when using the PHPUnit
+// phar due to some weird PHP behavior (the class gets defined without executing
+// the code before it and so the definition is not properly conditional)
+} else {
```
please revert the condition so we will not have empty `if` statement

keradus commented on src/Symfony/Bridge/PhpUnit/Legacy/CoverageListenerTrait.php:
```diff
@@ -0,0 +1,81 @@
+    {
+        $annotations = $test->getAnnotations();
+
+        if (isset($annotations['class']['covers']) || isset($annotations['method']['covers'])) {
```
add `@coversDefaultClass` and `@coversNothing`

keradus replied:
add `@coversDefaultClass` and `@coversNothing`

keradus commented on src/Symfony/Bridge/PhpUnit/Legacy/CoverageListenerTrait.php:
```diff
@@ -0,0 +1,81 @@
+ */
+class CoverageListenerTrait
+{
+    private $sutFqcnResolver;
```
could we please add typehints here and in other places?

lyrixx replied:
It's useless, it's already typehinted in the constructor.

And about `startTest` method: I did not find a wat because of PHPUnit namespace change. We have to support many version.

keradus replied:
> It's useless, it's already typehinted in the constructor.

constructor param is typehinted which is great, yet class property is not, thus when I want to use property in some method, I need to remember property method by looking at constructor, not from typehinting

lyrixx replied:
If you are using an IDE, The IDE do it for you.
If you don't (like me) the naming does everything.

keradus commented on src/Symfony/Bridge/PhpUnit/Legacy/CoverageListenerTrait.php:
```diff
@@ -0,0 +1,81 @@
+
+        $class = get_class($test);
+
+        $sutFqcn = str_replace('Tests\\', '', $class);
```
imo `'\\Tests\\', '\\'`

lyrixx replied:
Why not, but why ?

keradus replied:
to not have false positive on `FooTests` namespace, as we put tests in `Tests` subnamespace
I agree it's edge case, but let us be defensive

keradus replied:
would be easier to show such cases if we would have tests for this functoinality :(

lyrixx replied:
:+1:

keradus commented on src/Symfony/Bridge/PhpUnit/Legacy/CoverageListenerTrait.php:
```diff
@@ -0,0 +1,81 @@
+
+        $sutFqcn = $this->findSutFqcn($test);
+        if (!$sutFqcn) {
+            return;
```
Currently, If I have a scenario:
> Oh no, adding `@covers` annotation is boring, but I want to be strict about coverage and not have side-effect coverage reported, let me use this listener to automate it!

and then, when for some reason src class was not found (even typo in test class name like `Foo123` and `Foo124Test` (4 instead of 3)), one have no nice way to know about it.
Please, at least add a user warning (like we do in tests while using deprecated code)

lyrixx replied:
IMHO this is not a good idea. Because you may have some Integration / Functional / Acceptance tests. And theses test are not related to another class. It's just a place where to put some test.
So triggering a warning each time will be really boring for the end user.

keradus replied:
those would have `@coversNothing`.
so you will not hit this line for them

---------------------------------------------------------------

at least please put option to control this behaviour instead of silencing not-found scenario. as while using the listener you have no obvious way to see that src class was not found for given test class

lyrixx replied:
I'm ok to add an option to display something or not :+1:

keradus reviewed (changes requested):
Please, add tests so we will be sure it works like indented, especially across different phpunit verisons)

keradus commented on src/Symfony/Bridge/PhpUnit/Legacy/CoverageListenerTrait.php:
```diff
@@ -0,0 +1,81 @@
+{
+    private $sutFqcnResolver;
+
+    public function __construct(callable $sutFqcnResolver = null)
```
this class is 5.3 downgraded trait. how you gonna configure this with switching to real trait ?

lyrixx replied:
It will not be a trait but a normal class.

keradus replied:
then why it's called `CoverageListenerTrait` ?

`Trait` as suffix, then in description that it is intended to be a trait but low php version is not allowing it yet

nicolas-grekas replied:
the name is correct for consistency with existing trait-like classes

lyrixx replied:
To be consistant with other PHPUnit implementation ?!

keradus commented:
just to put focus to it again, PHPUnit is not developing really fast (new majors release quite often), and as @sebastianbergmann said:
> A test listener should only listen. In the future, it will only be able to listen.
> 
> The proper solution to this problem is to follow best practices and set forceCoversAnnotation="true" in your PHPUnit configuration.

lyrixx commented:
> A test listener should only listen. In the future, it will only be able to listen.

Yes I saw that. But there is no clean extension point :( We already ask to Sebastien how to do that in a clean manner, but he did not tell us how to.

keradus commented:
OK, so please add tests so we will be aware in nice way what's going on

lyrixx commented:
> OK, so please add tests so we will be aware in nice way what's going on

Yes I will try. You already ask ... let me some time please.

BTW, some others listener are not covered by tests either.

keradus commented:
that's bad for them.
while I'm not asking to fix everything at once, especially part you did not touch, there is no need to increase technical dept with new PRs (like increase not-tested code amount)

lyrixx committed:
Added test for the CoverageListener

lyrixx commented:
I have added some tests. But testing this Listener is not easy. What do you think of this test ?

lyrixx commented:
Arf, I don't know how to exclude my "Fixture" test from the Symfony test suite. I have try many things without success :/

lyrixx committed:
Added an option to throw a warning if the SUT solver can not find the SUT

lyrixx committed:
Added more tests + support more annotations

lyrixx committed:
Better FS structure & avoid to run fixture test in the main test suite

lyrixx committed:
Added a BC layer for older PHPUnit

lyrixx committed:
'Fixed' tests on HHVM

lyrixx committed:
Fixed tests on Windows

lyrixx committed:
Added note in CHANGELOG

lyrixx commented:
Hi.
So:

* I managed to fix test on travis & appveyor
* fabbot is wrong
* I addressed issue raised by @keradus 
* I added a note in the CHANGELOG

lyrixx commented:
https://github.com/symfony/symfony-docs/pull/8416

nicolas-grekas commented:
Still :+1: on my side, ping @symfony/deciders

dunglas commented:
👍

nicolas-grekas commented:
@lyrixx can you add an entry in the CHANGELOG file please?

lyrixx commented:
Oups, I forgot to push a commit. Fixed.

lyrixx committed:
Fixed issue with older PHPUnit Version

xabbuh commented on src/Symfony/Bridge/PhpUnit/CoverageListener.php:
```diff
@@ -0,0 +1,44 @@
+    class_alias('Symfony\Bridge\PhpUnit\Legacy\CoverageListener', 'Symfony\Bridge\PhpUnit\CoverageListener');
+// Using an early return instead of a else does not work when using the PHPUnit
+// phar due to some weird PHP behavior (the class gets defined without executing
+// the code before it and so the definition is not properly conditional)
```
fabbot does not seem to be happy with the comments not being indented

lyrixx replied:
It's a false positive. I ignored it.

fabpot replied:
How is that a false positive?

fabpot commented:
Thank you @lyrixx.
